### PR TITLE
More printf

### PR DESCRIPTION
**Pull request by dschwoerer**
We can tell gcc to find #1580 for us :+1: 

It did not compile - but I think that is #1572 

This probably should be made with a nice macro, and go into next?

I already spotted another bug, where the argument was ignored ... so probably should incorporate this one way or the other ...

### CODE REVIEW

File: include/bout/sys/expressionparser.hxx [hunk 1]
```diff
@@ -194,7 +194,8 @@ private:
 
 class ParseException : public std::exception {
 public:
-  ParseException(const char *, ...);
+  ParseException(const char *, ...)
+    __attribute__ ((format (printf, 2, 3)));
```
Comment (d7919) [inline include/bout/sys/expressionparser.hxx]: Isn't this `GCC` specific so may not work with any other compiler family?

Comment (ZedThree) [reply include/bout/sys/expressionparser.hxx]: Yes, this is non-portable, although it seems clang and Intel both accept it. It should be wrapped up in a macro.

One day we'll actually move to `{fmt}` and we'll get proper type-checking

Comment (dschwoerer) [reply include/bout/sys/expressionparser.hxx]: Probably, but if we can fix them at least with one compiler, than this is fine.

Also, gcc is available on most systems.

I tagged it as WIP, so we can guard it/put it in a macro if it causes issues for other compilers.

Comment (dschwoerer) [reply include/bout/sys/expressionparser.hxx]: thoughts on the macro name?
`_BOUT_FORMAT(i,j)?

Comment (ZedThree) [reply include/bout/sys/expressionparser.hxx]: We should avoid names with leading underscores followed by capitals. Maybe just `BOUT_FORMAT`, or `BOUT_FORMAT_ARGS`

Comment (dschwoerer) [reply include/bout/sys/expressionparser.hxx]: is master ok, or should it go into next only?

**Pull request merged**